xds/src/test/java/io/grpc/xds/XdsClientFederationTest.java

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.xds.XdsClient.ResourceWatcher;
+import io.grpc.xds.XdsListenerResource.LdsUpdate;
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Tests for xDS control plane federation scenarios.
+ */
+@RunWith(JUnit4.class)
+public class XdsClientFederationTest {
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockDirectPathWatcher;
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockWatcher;
+
+  private static final String SERVER_LISTENER_TEMPLATE_NO_REPLACEMENT =
+      "grpc/server?udpa.resource.listening_address=";
+
+  @Rule
+  public ControlPlaneRule trafficdirector = new ControlPlaneRule("test-server");
+
+  @Rule
+  public ControlPlaneRule directpathPa = new ControlPlaneRule(
+      "xdstp://server-one/envoy.config.listener.v3.Listener/test-server");
+
+  private XdsClient xdsClient;
+  private boolean originalFederationStatus;
+
+  @Before
+  public void setUp() throws XdsInitializationException {
+    MockitoAnnotations.initMocks(this);
+
+    originalFederationStatus = BootstrapperImpl.enableFederation;
+    BootstrapperImpl.enableFederation = true;
+
+    SharedXdsClientPoolProvider clientPoolProvider = new SharedXdsClientPoolProvider();
+    clientPoolProvider.setBootstrapOverride(defaultBootstrapOverride());
+    xdsClient = clientPoolProvider.getOrCreate().getObject();
+  }
+
+  @After
+  public void cleanUp() throws InterruptedException {
+    BootstrapperImpl.enableFederation = originalFederationStatus;
+  }
+
+  // Assures that resource deletions happening in one control plane do not trigger deletion events
+  /// in watchers of resources on other control planes.
+  @Test
+  public void isolatedResourceDeletions() throws InterruptedException {
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), "test-server", mockWatcher);
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(),
+        "xdstp://server-one/envoy.config.listener.v3.Listener/test-server", mockDirectPathWatcher);
+
+    trafficdirector.setLdsConfig(ControlPlaneRule.buildServerListener(),
+        ControlPlaneRule.buildClientListener("test-server"));
+    directpathPa.setLdsConfig(ControlPlaneRule.buildServerListener(),
+        ControlPlaneRule.buildClientListener(
+            "xdstp://server-one/envoy.config.listener.v3.Listener/test-server"));
+
+    // Deleting a resource (here by renaming it) in one control plan (here the "normal
```

Comment:
Is _deleting_ actually important? I'd have expected any change to the resource to be enough. I do see how it is easiest to delete the resource in the test, so that is fine, but the method comment and this comment seem to imply it is more specific.

Reply:
I tried to make the comment better. Note that the test control plane server is not fancy enough to store any state, it just passes whatever config it is given to the XdsClient as if no other config existed before. That's why the interaction with the test server isn't exactly like it would be with a real control plane.